### DISCUSSION TITLE
Issues Using Mikro-orm with Next.js and Vercel

### DISCUSSION THREAD

**Discussion by NebraskaCoder** [mikro-orm/mikro-orm#4411] [Q&A]
**Describe the bug**
I am unable to use mikro-orm in Next.js API route (serverless function) when hosted on Vercel. See the project at https://github.com/NebraskaCoder/reproduce-mikro-orm-vercel - when I try to go to {{url}}/api/tasks I get the error in the screenshot below. This is a copy of a private project with the bare minimum to show the problem. In the actual project, I am able to get data in getStaticProps without an issue, so this seems specific to serverless functions. In my config (at config/mikro-orm.ts), I have tried many different paths for `migrations.path`. The error never changes (even the paths in the error). I have tried removing `migrations {}` with no luck.

**Stack trace**
![Example Error](https://github.com/mikro-orm/mikro-orm/assets/64073162/0db595bc-940a-4911-9217-e57273c34eba)

**To Reproduce**
1. Clone my project at https://github.com/NebraskaCoder/reproduce-mikro-orm-vercel
2. Copy the .env.development.local.copy to .env.development.local
3. Define the env variables (at least POSTGRES_URL).

**Expected behavior**
I should not have an error when getting tasks.

**Versions**

| Dependency | Version |
| - | - |
| node | 18 |
| typescript | 5.0.4 |
| mikro-orm | 5.7.10 |
| @mikro-orm/postgresql | 5.7.10 |

**Comment by B4nan** [comment]
I am not sure if it's even feasible to use file-based migrations in serverless context, are you sure you can generally read/write files from inside it? 

Also not sure what exactly are you reporting. Your script cannot write into a folder, that sounds more like a vercel issue, do you really think this is about folder names? Note that the folder name is the same as in your config.

You could try to do it statically via the `migrationsList`, it was introduced (by community) for issues with webpack but this might be very similar:

https://mikro-orm.io/docs/next/migrations#importing-migrations-statically

Sidenote: you don't want to enable `allowGlobalContext`, that's gonna lead to weird issues almost certainly (unless your app instance is never reused, but I don't think that's the case with serverless). You want to have one context per request handler.

**Reply by NebraskaCoder** [reply]
Really, I don't want migrations on Vercel at all. If anything, I will be doing it locally or in CI/CD. I just need to get past the error that is preventing me from even using Mikro-orm. I wasn't aware that it was writing to anything. I figured it just needed to read it. I came to that conclusion when I removed the `await orm.getMigrator().up();` from the `lib/mikroorm.ts` file.

So... I just statically linked to the migrations via migrationsList, and that fixed my issue. I tried removing migrations from the config before and it still tried. VERCEL introduces an environmental variable (`VERCEL=1`) so you know your code is on Vercel. How can I disable migrations so I can check `process.env.VERCEL === 1` and disable it if the project is on Vercel?

Also on your sidenote: thank you for letting me know. I missed that. I will need to rework my code so it doesn't depend on it. It came up that way when making a global em. Should I be creating a new em instance as it is needed? I think the following code is what caused it:

```typescript
import {
  Connection,
  EntityManager,
  IDatabaseDriver,
  MikroORM,
} from "@mikro-orm/core";

import mikroOrmConfig from "@/config/mikro-orm";

import type { PostgreSqlDriver, SqlEntityManager } from "@mikro-orm/postgresql";

let entityManager:
  | (SqlEntityManager<PostgreSqlDriver> &
      EntityManager<IDatabaseDriver<Connection>>)
  | null = null;

export async function initMikroOrm(): Promise<EntityManager> {
  if (!entityManager) {
    const orm = await MikroORM.init<PostgreSqlDriver>(mikroOrmConfig);
    await orm.getMigrator().up(); // Optional: Run migrations if needed

    entityManager = orm.em;
  }

  return entityManager;
}
```

**Reply by NebraskaCoder** [reply]
Still the same question on my feedback but I solved the sidenote and updated the code (disabling allowGlobalContext).

**Reply by conspireagency** [reply]
@NebraskaCoder what worked for me was the NextJS + Mikro orm example linked in the docs, which works perfectly when you're working within API routes.  

In a couple instances I had to use functions in server rendered pages that access the database, and this is what I came up with as an extension of the demo repo solution... not sure it's optimal but perhaps @B4nan can weigh in.

using the "getEntityManager()" util function where I need it outside of a NextJS API route context ie. getServersideProps()

`import { MikroORM, RequestContext } from "@mikro-orm/core";
import { NextApiHandler, NextApiRequest, NextApiResponse } from "next";
import config from "../config/mikro-orm";
import { EntityManager } from "@mikro-orm/postgresql";

const getORM = async () => {
  if (!(global as any).__MikroORM__) {
    (global as any).__MikroORM__ = await MikroORM.init(config)
      // specific to in-memory sqlite
      .then(async (orm) => {
        //const generator = orm.getSchemaGenerator();
        //await generator.createSchema().catch();
        return orm;
      });
  }
  return (global as any).__MikroORM__;
};

const withORM =
  (handler: NextApiHandler) =>
  async (req: NextApiRequest, res: NextApiResponse) => {
    const orm = await getORM();
    return await RequestContext.create(orm.em, async () => handler(req, res));
  };

const getEM = () => {
  const em = RequestContext.getEntityManager();
  if (!em)
    throw new Error(
      "Entity manager not found. Are you in a 'withORM'-wrapped Context?"
    );
  return em;
};

const getEntityManager = async () => {
  const orm = await getORM();
  return orm.em.fork() as EntityManager;
};

export { withORM, getEM, getEntityManager };
`

Resisted serverless for some time now, but in some instances it makes life a lot easier.  TBH the example repo in the docs is pretty bare.  Feel like an example of using a function in getServersideProps would be a good addition.  Also, some notes on perhaps how to optimize connection pools for this use case 😄
